multiraft/multiraft.go

```diff
@@ -180,6 +181,55 @@ func (m *MultiRaft) sendEvent(event interface{}) {
 	}
 }
 
+// fanoutHeartbeat sends the given heartbeat to all groups which believe that
+// their leader resides on the sending node.
+func (s *state) fanoutHeartbeat(req *RaftMessageRequest) {
+	// A heartbeat message is expanded into a heartbeat for each group
+	// that the remote node is a part of.
+	originNode, _ := s.nodes[req.Message.From]
+	cnt := 0
+	for groupID := range originNode.groupIDs {
+		// If we don't think that the sending node is leading that group, don't
+		// propagate.
+		if s.groups[groupID].leader != req.Message.From || req.Message.From == s.nodeID {
+			log.V(8).Infof("node %v: not fanning out heartbeat to %v, msg is from %d and leader is %d",
+				s.nodeID, req.Message.To, req.Message.From, s.groups[groupID].leader)
+			continue
+		}
+		if err := s.multiNode.Step(context.Background(), groupID, req.Message); err != nil {
+			log.V(4).Infof("node %v: coalesced heartbeat step failed for message %s", s.nodeID, groupID,
```

Comment:
There aren't really any guidelines; it's pretty ad-hoc for now. One rule of thumb is that it should be reasonable to enable V(1) across all modules at the same time without getting too overwhelmed, whereas any higher-V logging should probably be turned on for one file at a time with --vmodule.